win32ss/user/ntuser/security.c

```diff
@@ -0,0 +1,493 @@
+                        &Buffer,
+                        &Size,
+                        MEM_RELEASE);
+}
+
+/**
+ * @brief
+ * Queries the authenticated user security identifier
+ * (SID) that is associated with the security context
+ * of the access token that is being opened.
+ *
+ * @param[out] User
+ * A pointer to the token user that contains the security
+ * identifier of the authenticated user.
+ *
+ * @return
+ * Returns STATUS_SUCCESS if the function has successfully
+ * queried the token user. STATUS_UNSUCCESSFUL is returned
+ * if the effective token of the caller couldn't be opened.
+ * STATUS_NO_MEMORY is returned if memory allocation for
+ * token user buffer has failed because of lack of necessary
+ * pages reserved for such allocation. A failure NTSTATUS
+ * code is returned otherwise.
+ *
+ * @remarks
+ * !!!WARNING!!! -- THE CALLER WHO QUERIES THE TOKEN USER IS
+ * RESPONSIBLE TO FREE THE ALLOCATED TOKEN USER BUFFER THAT IS
+ * BEING GIVEN.
+ */
+NTSTATUS
+IntQueryUserSecurityIdentification(
+    _Out_ PTOKEN_USER *User)
+{
+    NTSTATUS Status;
+    PTOKEN_USER UserToken;
+    HANDLE Token;
+    ULONG BufferLength;
+
+    /* Initialize the parameter */
+    *User = NULL;
+
+    /* Open the current token of the caller */
+    Token = IntGetCurrentAccessToken();
+    if (!Token)
+    {
+        ERR("IntQueryUserSecurityIdentification(): Couldn't capture the token!\n");
+        return STATUS_UNSUCCESSFUL;
+    }
+
+    /*
+     * Since we do not know what the length
+     * of the buffer size should be exactly to
+     * hold the user data, let the function
+     * tell us the size.
+     */
+    Status = ZwQueryInformationToken(Token,
+                                     TokenUser,
+                                     NULL,
+                                     0,
+                                     &BufferLength);
+    if (!NT_SUCCESS(Status) && Status == STATUS_BUFFER_TOO_SMALL)
+    {
+        /*
+         * Allocate some memory for the buffer
+         * based on the size that the function
+         * gave us.
+         */
+        UserToken = IntAllocateSecurityBuffer(BufferLength);
+        if (!UserToken)
+        {
+            /* Bail out if we failed */
+            ERR("IntQueryUserSecurityIdentification(): Couldn't allocate memory for the token user!\n");
+            ZwClose(Token);
+            return STATUS_NO_MEMORY;
+        }
+    }
+
+    /* Query the user now as we have plenty of space to hold it */
+    Status = ZwQueryInformationToken(Token,
+                                     TokenUser,
+                                     UserToken,
+                                     BufferLength,
+                                     &BufferLength);
+    if (!NT_SUCCESS(Status))
+    {
+        /* We failed, bail out */
+        ERR("IntQueryUserSecurityIdentification(): Failed to query token user (Status 0x%08lx)\n", Status);
+        ZwClose(Token);
+        IntFreeSecurityBuffer(UserToken);
```

Comment:
unwinding in reverse order of init:
```suggestion
        IntFreeSecurityBuffer(UserToken);
        ZwClose(Token);
```